src/AggregateFunctions/AggregateFunctionMinMax.h

```diff
@@ -0,0 +1,1052 @@
+    static ALWAYS_INLINE inline bool column_compareAt(const IColumn & column, size_t n, size_t m)
+    {
+        return column.compareAt(n, m, column, /* nan/null direction hint = */ 1) > 0;
+    }
+#if USE_EMBEDDED_COMPILER
+    static ALWAYS_INLINE inline llvm::Value * llvm_cmp(llvm::IRBuilder<> & b, llvm::Value * value_to_check, llvm::Value * value)
+    {
+        if constexpr (!canBeNativeType<T>())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Unsupported type");
+        if constexpr (std::is_floating_point_v<Type>)
+            return b.CreateFCmpOGT(value_to_check, value);
+        if constexpr (std::numeric_limits<Type>::is_signed)
+            return b.CreateICmpSGT(value_to_check, value);
+        else
+            return b.CreateICmpUGT(value_to_check, value);
+    }
+#endif
+};
+
+template <typename Comparator>
+struct AggregateFunctionsMinMaxDataNumeric
+{
+    using Self = AggregateFunctionsMinMaxDataNumeric<Comparator>;
+    using ComparatorType = Comparator;
+    using T = typename Comparator::Type;
+    using ColVecType = ColumnVectorOrDecimal<T>;
+    /// We want to use multitarget code and dynamic dispatch only when it makes things faster, which is basically with native types
+    static constexpr bool supported_multitarget_type = std::is_integral_v<T> || std::is_floating_point_v<T>;
+    static constexpr bool allocates_memory_in_arena = false;
+
+    bool has_value = false;
+    T value{};
+
+    bool ALWAYS_INLINE has() const { return has_value; }
+
+    bool ALWAYS_INLINE inline add(const T & new_value)
+    {
+        if (!has_value || Comparator::cmp(new_value, value))
+        {
+            value = new_value;
+            has_value = true;
+            return true;
+        }
+        return false;
+    }
+
+    bool ALWAYS_INLINE inline add(const IColumn & column, size_t row_num, Arena *)
+    {
+        const auto & col = assert_cast<const ColVecType &>(column);
+        return add(col.getData()[row_num]);
+    }
+
+    void ALWAYS_INLINE set(const IColumn & column, size_t row_num, Arena *)
+    {
+        const auto & col = assert_cast<const ColVecType &>(column);
+        value = col.getData()[row_num];
+        has_value = true;
+    }
+
+    void ALWAYS_INLINE inline set(const AggregateFunctionsMinMaxDataNumeric & rhs, Arena *)
+    {
+        has_value = rhs.has_value;
+        value = rhs.value;
+    }
+
+    bool ALWAYS_INLINE inline addManyDefaults(const IColumn & column, size_t /*length*/, Arena * arena) { return add(column, 0, arena); }
+
+    void insertResultInto(IColumn & to) const
+    {
+        if (has())
+            assert_cast<ColVecType &>(to).getData().push_back(value);
+        else
+            assert_cast<ColVecType &>(to).insertDefault();
+    }
+
+    MULTITARGET_FUNCTION_AVX2_SSE42(
+    MULTITARGET_FUNCTION_HEADER(template <bool add_all_elements, bool add_if_cond_zero> static std::optional<T> NO_INLINE),
+    findNumericExtremeImpl,
```

Comment:
Why could not we add this optimized methods for native numeric types in previous `SingleValueDataFixed` and in functions Min, Max, ArgMin, ArgMax use them ?

Reply:
We could. But it would have required changing the APIs of all aggregations at once and that's really too much to do in one go. The current API is not great to introduce batch processing, so it was simpler to introduce batch processing to one set of functions first, then work on adapting the other set of functions (ANY*) to the new API and finally, if possible, merge them back together.

Reply:
It does not look to me that you have to change everything at once.
For example lets try to optimize `min` function first. You can add your optimized methods in `AggregateFunctionMinData`. Then in `AggregateFunctionsSingleValue` in `addBatchSinglePlace` check if `Data` is `AggregateFunctionMinData` and use this specialized method. The same can be implemented for `AggregateFunctionMaxData`, and probably `AggregateFunctionMinData` and `AggregateFunctionMaxData` could share some implementation details.
Is there any hidden issues, what do you think ?

Reply:
The last time I tried doing something like this I ended up having to add a hack:

https://github.com/ClickHouse/ClickHouse/blob/df3e4d26ad4506946cfd19eebef18d2bb1f208bd/src/AggregateFunctions/AggregateFunctionMinMaxAny.h#L1187-L1189

I had to do this because `AggregateFunctionXXXData` are themselves templates and I couldn't find [a way to do it](https://github.com/ClickHouse/ClickHouse/pull/34760#discussion_r810607602).

Also, this `changeIfBetter` API is unique to these functions. Other aggregates are using `add`, `addMany` and so on that I find more intuitive. I also prefer having different classes for every different aggregate function; you can have the same base class and specially add() instead of having `changeFirstTime`, `changeEveryTime`, `changeIfGreater`, etc all in the same place.

Try to keep in mind that they currently share a lot of the code because the code was made in such a way to be shared. For instance, why do you need a `compileChangeEveryTime`? It doesn't make sense to have an specialization of this when what it should be doing is iterate for the end until it finds a not-null value and insert it, not replace the value times again and again (but faster because of JIT). 

Finally, I'd say that it takes quite a bit of time to try to improve anything in these functions because a single change requires rebuilding all associated aggregate functions (super slow), and then since any could have changed you need to test them all and then the surface of changes keeps growing and conflicting with other PRs. To me the best way to approach this was to first break them apart, but maybe other people find it simpler to do it the other way.

Reply:
Seems reasonable. If you have a better approach feel free to integrate the code differently.